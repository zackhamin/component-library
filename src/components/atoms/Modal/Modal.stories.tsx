import { useState } from 'react';
import { Modal, ModalProps } from './Modal';

export default {
	title: 'atoms/Modal',
	component: Modal,
};

export const ModalStory = (args: ModalProps) => {
	const [isOpen, setIsOpen] = useState(true);

	const toggleOpen = () => setIsOpen((prev) => !prev);

	return (
		<div
			style={{
				width: '100vw',
				height: '100vh',
				background: 'url(https://source.unsplash.com/random)',
				backgroundSize: 'cover',
			}}
		>
			<button onClick={toggleOpen}>Open Modal</button>
			{isOpen && <Modal {...args} closeHandler={toggleOpen} />}
		</div>
	);
};

ModalStory.storyName = 'Modal';
ModalStory.parameters = {
	layout: 'fullscreen',
};
ModalStory.args = {
	title: 'My Modal',
	children: (
		<p>
			Lorem ipsum, dolor sit amet consectetur adipisicing elit. Consequuntur
			doloribus alias quis iste amet! Neque molestias itaque officia at nam
			quaerat ipsam velit quisquam, deserunt, eveniet quidem sint, pariatur vel.
		</p>
	),
};
ModalStory.argTypes = {
	closeHandler: { action: 'closed' },
};
