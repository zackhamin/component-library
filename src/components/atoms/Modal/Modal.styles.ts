import { colors } from 'Styles';
import styled from 'styled-components';

const StyledBackdrop = styled.div`
	display: flex;
	position: fixed;
	width: 100vw;
	height: 100vh;
	top: 0;
	left: 0;
	backdrop-filter: blur(1rem);
	background-color: rgba(255, 255, 255, 0.5);
	justify-content: center;
	align-items: center;
	z-index: 1;
`;

const StyledModal = styled.div`
	background: ${colors.white};
	border-radius: 0.8rem;
	box-shadow: 0px 2px 6px rgb(51 51 51 / 16%);
	width: 90rem;
	max-width: 90vw;
	min-height: 30rem;
	max-height: 90vh;
	overflow: scroll;
	padding: 2.4rem;
	position: relative;
	z-index: 2;

	button.close {
		color: ${colors.emerald};
		font-size: 2.4rem;
		padding: 0;
		margin: 0;
		border: 0;
		background: transparent;
		position: absolute;
		top: 0.8rem;
		right: 0.8rem;
		z-index: 3;
	}
`;

export { StyledModal, StyledBackdrop };
