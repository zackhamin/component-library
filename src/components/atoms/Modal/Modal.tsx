import FocusTrap from 'focus-trap-react';
import { Heading } from '..';
import { IconClose } from 'Atoms/Icons';
import { useOnKeyDown } from 'Hooks/useOnKeyDown';
import { ReactNode, useEffect } from 'react';
import { StyledBackdrop, StyledModal } from './Modal.styles';

export interface ModalProps {
	children: ReactNode;
	title: string;
	closeHandler: () => void;
}

const Modal = ({ children, closeHandler, title }: ModalProps) => {
	useOnKeyDown('Escape', closeHandler);

	useEffect(() => {
		document.body.style.overflow = 'hidden';

		return () => {
			document.body.style.overflow = 'unset';
		};
	}, []);

	return (
		<StyledBackdrop onClick={closeHandler} data-testid="backdrop">
			<FocusTrap
				focusTrapOptions={{
					allowOutsideClick: true,
					delayInitialFocus: false,
				}}
			>
				<StyledModal
					role="dialog"
					aria-modal="true"
					aria-labelledby="modal-title"
					onClick={(e) => e.stopPropagation()}
					tabIndex={-1}
				>
					<button
						className="close"
						aria-label="Close"
						onClick={closeHandler}
						data-testid="close"
					>
						<IconClose />
					</button>
					<Heading id="modal-title">{title}</Heading>
					{children}
				</StyledModal>
			</FocusTrap>
		</StyledBackdrop>
	);
};

export { Modal };
