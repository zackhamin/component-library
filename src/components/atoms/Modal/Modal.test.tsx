import { Modal } from '../Modal';
import { fireEvent, render } from '@testing-library/react';

jest.mock('focus-trap-react', () => ({
	__esModule: true,
	default: ({ children }: any) => <div id="focus-trap">{children}</div>,
}));

describe('Modal', () => {
	it('renders correctly', () => {
		const { container } = render(
			<Modal title="My Modal" closeHandler={jest.fn}>
				Modal Content
			</Modal>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('calls close handler on click of backdrop and close button', () => {
		const closeHandler = jest.fn();
		const { getByTestId } = render(
			<Modal title="My Modal" closeHandler={closeHandler}>
				Modal Content
			</Modal>
		);

		const backdrop = getByTestId('backdrop');
		const closeButton = getByTestId('close');

		fireEvent.click(backdrop);
		fireEvent.click(closeButton);
		expect(closeHandler).toHaveBeenCalledTimes(2);
	});
});
