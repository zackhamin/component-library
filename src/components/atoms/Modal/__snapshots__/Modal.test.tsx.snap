// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
.c2 {
  margin: 0;
  font-family: BrandonGrotesque-Bold,Arial,Helvetica,sans-serif;
  color: #333333;
  font-size: 2.6rem;
  line-height: 140%;
  font-weight: 450;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  width: 100vw;
  height: 100vh;
  top: 0;
  left: 0;
  -webkit-backdrop-filter: blur(1rem);
  backdrop-filter: blur(1rem);
  background-color: rgba(255,255,255,0.5);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 1;
}

.c1 {
  background: #ffffff;
  border-radius: 0.8rem;
  box-shadow: 0px 2px 6px rgb(51 51 51 / 16%);
  width: 90rem;
  max-width: 90vw;
  min-height: 30rem;
  max-height: 90vh;
  overflow: scroll;
  padding: 2.4rem;
  position: relative;
  z-index: 2;
}

.c1 button.close {
  color: #00594C;
  font-size: 2.4rem;
  padding: 0;
  margin: 0;
  border: 0;
  background: transparent;
  position: absolute;
  top: 0.8rem;
  right: 0.8rem;
  z-index: 3;
}

<div
  class="c0"
  data-testid="backdrop"
>
  <div
    id="focus-trap"
  >
    <div
      aria-labelledby="modal-title"
      aria-modal="true"
      class="c1"
      role="dialog"
      tabindex="-1"
    >
      <button
        aria-label="Close"
        class="close"
        data-testid="close"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.707 6.707a1 1 0 0 0-1.414-1.414L12 10.586 6.707 5.293a1 1 0 0 0-1.414 1.414L10.586 12l-5.293 5.293a1 1 0 1 0 1.414 1.414L12 13.414l5.293 5.293a1 1 0 0 0 1.414-1.414L13.414 12l5.293-5.293z"
            fill="currentColor"
          />
        </svg>
      </button>
      <h2
        class="c2"
        id="modal-title"
      >
        My Modal
      </h2>
      Modal Content
    </div>
  </div>
</div>
`;
